RN-175: Add order columns transform (p1. switch to managing columns explicitly in report-server)

rohan-bes opened the pull request:
### Issue RN-175:

Part 1 of 3.

We want to introduce an ability to order the columns of the data in the viz-builder. However, before we do this, we need to actually have an explicit order to the columns in the report-server in the first place.

There are other reasons it is advantageous to switch to tracking an explicit column order as well:
- It opens opportunities for controlling how vizes are presented more easily
- It allows us to introduce excel style columns range lookups (eg. `sum(col1:col5)`)
- It makes behaviour of transforms a bit more consistent

Here we are introducing a simple data-structure called a `TransformTable`. It manages the rows of the table as before, but also separately tracks an array of columns. It has a number of utility methods for manipulating the rows/columns atomically, so that the caller doesn't need to ensure they're kept in sync manually.

Tried to make this review as digestable as possible for the reviewer by breaking up changes to each transform step into its own commit, so going commit-by-commit might be the easiest approach.

billli0 commented on packages/report-server/src/__tests__/reportBuilder/transform/mergeRows.test.ts:
```diff
@@ -46,7 +46,11 @@ describe('mergeRows', () => {
       },
     ]);
-    expect(transform(MERGEABLE_ANALYTICS)).toEqual([{ BCD1: 28, BCD2: 123 }]);
+    expect(transform(TransformTable.fromRows(MERGEABLE_ANALYTICS))).toEqual(
+      TransformTable.fromRows([
+        { period: undefined, organisationUnit: undefined, BCD1: 28, BCD2: 123 },
+      ]),
+    );
   });
 
```
As we have undefined value, suggest to use `toStrictEqual` instead of `toEqual`. Difference: https://jestjs.io/docs/expect#tostrictequalvalue

Also I think we should use `toStrictEqual` in other unit tests to ensure that transforms don't add unexpected undefined value. 

As we are refactoring all unit tests in this PR, should be a good chance to switch?

rohan-bes replied:
In this specific case, I added `undefined` values more to just get the columns to appear in the correct order then to validate the data.

I think I'm generally a bit torn on using `toStrictEqual`. It would be more accurate, but it might make the tests more painful to write/maintain if there are frequent `undefined` values (which I suspect there might be...)

I'll try switching and see how things go 👍

billli0 replied:
Thanks for switching that🙏

billli0 commented on packages/report-server/src/reportBuilder/transform/aliases/keyValueByFieldAliases.ts:
```diff
@@ -3,37 +3,56 @@
+    newColumnNames.map(columnName => [
+      columnName,
+      // Flag all values as skip char to avoid overwriting data
+      table.hasColumn(columnName)
```
Should `table.hasColumn(columnName)` always be true? `newColumnNames` are built from `table`.

billli0 commented on packages/report-server/src/reportBuilder/transform/aliases/keyValueByFieldAliases.ts:
```diff
@@ -3,37 +3,56 @@
+import { TransformTable } from '../table';
+import { FieldValue } from '../../types';
+
+const keyValueByField = (table: TransformTable, field: string) => {
+  const fieldValues = table.getColumnValues(field) as string[];
+  const valueValues = table.getColumnValues('value');
+  const newColumnNames = Array.from(new Set(fieldValues));
+  const newColumns: Record<string, FieldValue[]> = Object.fromEntries(
+    newColumnNames.map(columnName => [
+      columnName,
```
Do you think it would be simpler to use the logic `get Rows -> modify Rows -> build table from Rows`?

rohan-bes replied:
Hmmm yeah probably would be, I'll switch it it work that way 👍

billli0 commented on packages/report-server/src/reportBuilder/transform/functions/updateColumns.ts:
```diff
@@ -28,33 +29,54 @@ export const paramsValidator = yup.object().shape({
-  return rows.map(row => {
-    const returnNewRow = params.where(parser);
-    if (!returnNewRow) {
+const updateColumns = (table: TransformTable, params: UpdateColumnsParams, context: Context) => {
+  const parser = new TransformParser(table, context);
+  const newColumns: Record<string, FieldValue[]> = {};
+  const skippedRows: Record<number, Row> = {};
+  const columnNames = table.getColumns();
```
Suggest to clone table first, so the variable `table` is immuable.

rohan-bes replied:
`getColumns` actually returns a new array of the column names. So mutating those columns won't mutate the original table 👍

rohan-bes commented:
Sorry @billli0 I merged the other part PRs before noticing that you wanted to re-review this one 🤦 

I addressed the issues you mentioned in these two commits of you wanna review them?
- Make tests use strictEquals: https://github.com/beyondessential/tupaia/pull/4191/commits/9278f74769703d8c0e19e0b727e4b84c2cec0723
- Use simpler logic for keyBy aliases:  https://github.com/beyondessential/tupaia/pull/4191/commits/7c0ac43119d4857e76f268d4316bc53c819d1cc3
